compiler/aarch64/codegen/ARM64BinaryEncoding.cpp

```diff
@@ -534,3 +535,71 @@ uint8_t *TR::ARM64Src2Instruction::generateBinaryEncoding()
    setBinaryEncoding(instructionStart);
    return cursor;
    }
+
+#ifdef J9_PROJECT_SPECIFIC
+uint8_t *TR::ARM64VirtualGuardNOPInstruction::generateBinaryEncoding()
+   {
+   uint8_t *cursor = cg()->getBinaryBufferCursor();
+   TR::LabelSymbol *label = getLabelSymbol();
+   int32_t length = 0;
+   TR::Instruction *guardForPatching = cg()->getVirtualGuardForPatching(this);
+
+   // a previous guard is patching to the same destination and we can recycle the patch
+   // point so setup the patching location to use this previous guard and generate no
+   // instructions ourselves
+   if ((guardForPatching != this) &&
+         // AOT needs an explicit nop, even if there are patchable instructions at this site because
+         // 1) Those instructions might have AOT data relocations (and therefore will be incorrectly patched again)
+         // 2) We might want to re-enable the code path and unpatch, in which case we would have to know what the old instruction was
+         !cg()->comp()->compileRelocatableCode())
+      {
+      _site->setLocation(guardForPatching->getBinaryEncoding());
+      setBinaryLength(0);
+      setBinaryEncoding(cursor);
+      if (label->getCodeLocation() == NULL)
+         {
+         cg()->addRelocation(new (cg()->trHeapMemory()) TR::LabelAbsoluteRelocation((uint8_t *) (&_site->getDestination()), label));
+         }
+      else
+         {
+         _site->setDestination(label->getCodeLocation());
+         }
+      cg()->addAccumulatedInstructionLengthError(getEstimatedBinaryLength() - getBinaryLength());
+      return cursor;
+      }
+
```

Comment:
I think we need to leave a comment here to revisit this. While the implementation is fine we need to improve it to support empty patching. See example on Z [1]. There is also one in x86 too I believe.

[1] https://github.com/eclipse/omr/blob/cf9d75a4a51dfa49103682784481b461777a2d41/compiler/z/codegen/S390Instruction.cpp#L5560-L5572

Reply:
I added a comment to revisit this. Thanks for the example.